File: leetcode/22.Generate Parentheses.cpp
/*
 * @lc app=leetcode.cn id=22 lang=cpp
 *
 * [22] 括号生成
 */

// @lc code=start

#include <string>
#include <vector>

using namespace std;

class Solution {
   public:
    void helper(vector<string>& result, string str, int& n, int left,
                int right) {
        if(left - right > 0){
            if(left  < n) {
                helper(result, str + '(', n, left + 1, right);
            }
            helper(result, str + ')', n, left, right + 1);
        }else {
            if(right < n){
                helper(result, str + '(', n, left + 1, right);
            }else {
                result.emplace_back(str);
                return;
            }
        }
    }

    vector<string> generateParenthesis(int n) {
        if (n <= 0) {
            return {""};
        }
        string str = "(";
        vector<string> result;

        helper(result, str, n, 1, 0);

        return result;
    }
};
// @lc code=end


int main(){
    Solution s;
    s.generateParenthesis(3);
    return 0;
}

File: leetcode/_22_Generate_Parentheses.java
import java.util.ArrayList;
import java.util.List;

public class _22_Generate_Parentheses {
    class Solution {
        public List<String> generateParenthesis(int n) {
            List<String> results = new ArrayList<>();
            dfs(results, new ArrayList<>(), 0, 0, n);
            return results;
        }

        private void dfs(List<String> results, List<String> result, int left, int right, int n) {
            if (left + right == n * 2) {
                results.add(String.join("", result));
                return;
            }
            if (left < n) {
                result.add("(");
                dfs(results, result, left + 1, right, n);
                result.remove(result.size() - 1);
            }
            if (left > right) {
                result.add(")");
                dfs(results, result, left, right + 1, n);
                result.remove(result.size() - 1);
            }
        }
    }
}